src/worker.cc

```diff
@@ -1233,12 +1089,93 @@ worker_t::result_t to_response(const std::string& data,
 
 #endif
 
-service_worker_t::service_worker_t() : interrupt(nullptr) {
+// TODO: when we want to use this in mjolnir too we can move this into a private header
+// this is a wrapper of a third party lib that provides a client for statsd integration
+// since metrics are important both for on- and offline processing we keep the impl here
+// running services can use it and also data ETL that lives in mjolnir can use it
+struct statsd_client_t : public Statsd::StatsdClient {
+  statsd_client_t(const boost::property_tree::ptree& conf)
+      : Statsd::StatsdClient(conf.get<std::string>("statsd.host", ""),
+                             conf.get<int>("statsd.port", 8125),
+                             conf.get<std::string>("statsd.prefix", ""),
+                             conf.get<uint64_t>("statsd.batch_size", 500),
+                             0) {
+    auto added_tags = conf.get_child_optional("statsd.tags");
+    if (added_tags) {
+      for (const auto& tag : *added_tags) {
+        tags.push_back(tag.second.data());
+      }
+    }
+  }
+  std::vector<std::string> tags;
+};
+
+service_worker_t::service_worker_t(const boost::property_tree::ptree& config)
+    : interrupt(nullptr), statsd_client(new statsd_client_t(config)) {
 }
 service_worker_t::~service_worker_t() {
 }
 void service_worker_t::set_interrupt(const std::function<void()>* interrupt_function) {
   interrupt = interrupt_function;
 }
+void service_worker_t::cleanup() {
+  // sends metrics to statsd server over udp
+  statsd_client->flush();
```

Comment:
this is what actually sends the metrics to statsd. notice that its in the cleanup method so it doesnt block the request flow its specifically done when the worker is not actively processing a request

Reply:
And this will result in multiple sends if the total batched payload is larger than the configured batch-size?

Reply:
correct, more information is available in the library documentation: https://github.com/vthiery/cpp-statsd-client#batching